stdlib/universe/difference_test.go

```diff
@@ -723,3 +723,242 @@ func TestDifference_Process(t *testing.T) {
+			spec: &universe.DifferenceProcedureSpec{
+				Columns:     []string{execute.DefaultValueColLabel},
+				NonNegative: true,
+				InitialZero: true,
+			},
+			data: []flux.Table{&executetest.Table{
+				ColMeta: []flux.ColMeta{
+					{Label: "_time", Type: flux.TTime},
+					{Label: "_value", Type: flux.TFloat},
+				},
+				Data: [][]interface{}{
+					{execute.Time(1), 20.0},
+					{execute.Time(2), 30.0},
+					{execute.Time(3), 50.0},
+					{execute.Time(4), 30.0},
+					{execute.Time(5), -40.0},
+				},
+			}},
+			want: []*executetest.Table{{
+				ColMeta: []flux.ColMeta{
+					{Label: "_time", Type: flux.TTime},
+					{Label: "_value", Type: flux.TFloat},
+				},
+				Data: [][]interface{}{
+					{execute.Time(2), 10.0},
+					{execute.Time(3), 20.0},
+					{execute.Time(4), 30.0},
+					{execute.Time(5), nil},
+				},
+			}},
+		},
+		// keepFirst: true, initialZero: true
+		{
+			name: "float with KeepFirst and InitialZero",
+			spec: &universe.DifferenceProcedureSpec{
+				Columns:     []string{execute.DefaultValueColLabel},
+				NonNegative: true,
+				KeepFirst:   true,
+				InitialZero: true,
+			},
+			data: []flux.Table{&executetest.Table{
+				ColMeta: []flux.ColMeta{
+					{Label: "_time", Type: flux.TTime},
+					{Label: "_value", Type: flux.TFloat},
+					{Label: "t", Type: flux.TString},
+				},
+				Data: [][]interface{}{
+					{execute.Time(1), 20.0, "a"},
+					{execute.Time(2), 30.0, "b"},
+					{execute.Time(3), 50.0, "c"},
+					{execute.Time(4), 30.0, "d"},
+					{execute.Time(5), 40.0, "e"},
+				},
+			}},
+			want: []*executetest.Table{{
+				ColMeta: []flux.ColMeta{
+					{Label: "_time", Type: flux.TTime},
+					{Label: "_value", Type: flux.TFloat},
+					{Label: "t", Type: flux.TString},
+				},
+				Data: [][]interface{}{
+					{execute.Time(1), float64(0), "a"},
+					{execute.Time(2), 10.0, "b"},
+					{execute.Time(3), 20.0, "c"},
+					{execute.Time(4), 30.0, "d"},
+					{execute.Time(5), 10.0, "e"},
+				},
+			}},
+		},
+		{
+			name: "with null non negative and InitialZero",
+			spec: &universe.DifferenceProcedureSpec{
+				Columns:     []string{"a", "b", "c"},
+				NonNegative: true,
+				InitialZero: true,
+			},
+			data: []flux.Table{&executetest.Table{
+				ColMeta: []flux.ColMeta{
+					{Label: "_time", Type: flux.TTime},
+					{Label: "a", Type: flux.TFloat},
+					{Label: "b", Type: flux.TFloat},
+					{Label: "c", Type: flux.TFloat},
+				},
+				Data: [][]interface{}{
+					{execute.Time(0), nil, 1.0, 2.0},
+					{execute.Time(1), 6.0, 2.0, nil},
+					{execute.Time(2), 4.0, 2.0, 4.0},
+					{execute.Time(3), 10.0, 10.0, 2.0},
+					{execute.Time(4), nil, nil, 1.0},
+				},
+			}},
+			want: []*executetest.Table{{
+				ColMeta: []flux.ColMeta{
+					{Label: "_time", Type: flux.TTime},
+					{Label: "a", Type: flux.TFloat},
+					{Label: "b", Type: flux.TFloat},
+					{Label: "c", Type: flux.TFloat},
+				},
+				Data: [][]interface{}{
+					{execute.Time(1), nil, 1.0, nil},
```

Comment:
My understanding is that this should have `a = 6.0`? If it's initial zero, then the difference is seeded at zero. We encounter a null so we would emit a null if `keepFirst` was true, but since `keepFirst` is false we swallow it (hence not in the output). Then we see time 1, see a 6, and `6 - 0 = 6`.

Reply:
we are not seeding difference to zero for the case `(keepFirst: false, initialZero: true)`. when both of them are true then only 0 value is returned. https://github.com/influxdata/flux/blob/cda99ce8b5c527710b34389b74d853b53c35bea2/stdlib/universe/difference.go#L410

The current behaviour is similar to existing [difference](https://docs.influxdata.com/flux/v0.x/stdlib/universe/difference/) which is 
`Some value v minus null is v minus the last non-null value seen before v; "or null if v is the first non-null value seen".`

Reply:
That's my point. When initial zero is true, the last non-null value is always zero. That makes me think that, to be consistent, this should output 6.

Reply:
Never mind on this. I think I got myself confused. We should not seed the first value with zero as that can create a large bump at the beginning. I'll rereview with that in mind.